pkg/sensors/accelerator/accelerator.go

```diff
@@ -0,0 +1,231 @@
+			m:        sync.Mutex{},
+			Registry: map[device.DeviceType]Accelerator{},
+		}
+	})
+	return globalRegistry
+}
+
+// SetRegistry replaces the global registry instance
+// NOTE: All plugins will need to be manually registered
+// after this function is called.
+func SetRegistry(registry *AcceleratorRegistry) {
+	globalRegistry = registry
+}
+
+// Devices returns a map of supported accelerators.
+func (r *AcceleratorRegistry) Accelerators() (map[device.DeviceType]Accelerator, error) {
+	if len(r.Registry) == 0 {
+		return nil, errors.New("no accelerators found")
+	}
+	return r.Registry, nil
+}
+
+// ActiveAcceleratorsByType returns a map of supported accelerators based on the specified type...
+func (r *AcceleratorRegistry) ActiveAcceleratorsByType(t AcceleratorType) (map[device.DeviceType]Accelerator, error) {
+	acc := map[device.DeviceType]Accelerator{}
+	for _, a := range r.Registry {
+		if a.AccType() == t && a.IsRunning() {
+			d := a.Device()
+			if d.IsDeviceCollectionSupported() {
+				acc[d.DevType()] = a
+			}
+		}
+	}
+	if len(acc) == 0 {
+		return nil, errors.New("accelerators not found")
+	}
+	return acc, nil
+}
+
+func CreateAndRegister(atype AcceleratorType, r *AcceleratorRegistry, sleep bool) error {
+	var numDevs int
+	var getDevices func() []device.DeviceType
+	maxDeviceInitRetry := 10
+
+	switch atype {
+	case GPU:
+		numDevs = len(device.GetGpuDevices())
+		getDevices = device.GetGpuDevices // returns a slice of registered GPU devices[NVML|DCGM|HABANA]. TODO CHECK IF WE NEED TO MAKE THIS SINGULAR
+	case DUMMY:
+		numDevs = len(device.GetDummyDevices())
+		getDevices = device.GetDummyDevices
+	default:
+		return errors.New("unsupported accelerator")
+	}
+
+	if numDevs != 0 {
+		devices := getDevices()
+
+		klog.V(5).Infof("Initializing the %s Accelerator collectors in type %v", atype.String(), devices)
+
+		for _, devType := range devices {
+			for i := 0; i <= maxDeviceInitRetry; i++ {
+				// Create a new accelerator and add it to the registry
+				if err := NewAccelerator(devType, atype, r); err != nil {
+					klog.Errorf("Could not init the %s Accelerator collector going to try again", devType.String())
+					if sleep {
+						// The GPU operators typically takes longer time to initialize than kepler resulting in error to start the gpu driver
+						// therefore, we wait up to 1 min to allow the gpu operator initialize
+						time.Sleep(6 * time.Second)
+					}
+					continue
+				}
+				klog.V(5).Infof("Startup %s Accelerator collector successful", devType.String())
+				break
+			}
+		}
+	} else {
+		return errors.New("No Accelerator collectors found")
+	}
+	return nil
+}
+
+// NewAccelerator creates a new Accelerator instance with a specific device [NVML|DCGM|DUMMY|HABANA] for the local node.
+func NewAccelerator(devType device.DeviceType, atype AcceleratorType, r *AcceleratorRegistry) error {
+	var d device.DeviceInterface
+	var err error
+
+	_, ok := r.Registry[devType] // e.g. accelerators[nvml|dcgm|habana|dummy]
```

Comment:
Shouldn't you mutex lock for read? I think sync.Map may be a better candidate if we feel registry will be accessed by multiple threads.